[CI] ReloadingDatabasesWhilePerformingGeoLookupsIT#test() failure

**Build scan**: https://gradle-enterprise.elastic.co/s/3wd2py7tycbjm

**Repro line**: `./gradlew ':modules:ingest-geoip:internalClusterTest' --tests "org.elasticsearch.ingest.geoip.ReloadingDatabasesWhilePerformingGeoLookupsIT.test" -Dtests.seed=1327A9992993281D -Dtests.security.manager=true -Dtests.locale=es-GT -Dtests.timezone=Indian/Chagos -Druntime.java=11`

**Reproduces locally?**: Yes (when running with `-Dtests.iters=1024`, the test seed is irrelevant, it is a timing issue)

**Applicable branches**: master

**Failure history**: https://gradle-enterprise.elastic.co/scans/tests?search.relativeStartTime=P7D&search.timeZoneId=Europe/Amsterdam&tests.container=org.elasticsearch.ingest.geoip.ReloadingDatabasesWhilePerformingGeoLookupsIT&tests.sortField=FAILED&tests.test=test&tests.unstableOnly=true

**Failure excerpt**: 

```

java.lang.AssertionError: |  
-- | --
  | Expected: null |  
  | but: was <java.lang.AssertionError> |  

  |   | at __randomizedtesting.SeedInfo.seed([1327A9992993281D:9B739643876F45E5]:0) |  
-- | -- | -- | --
  |   | at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18) |  
  |   | at org.junit.Assert.assertThat(Assert.java:956) |  
  |   | at org.junit.Assert.assertThat(Assert.java:923) |  
  |   | at org.elasticsearch.ingest.geoip.ReloadingDatabasesWhilePerformingGeoLookupsIT.test(ReloadingDatabasesWhilePerformingGeoLookupsIT.java:142)


```

This is the actual underlying issue:

```
java.lang.AssertionError: null
	at org.elasticsearch.ingest.geoip.DatabaseRegistry.removeStaleEntries(DatabaseRegistry.java:258) ~[main/:?]
	at org.elasticsearch.ingest.geoip.ReloadingDatabasesWhilePerformingGeoLookupsIT.lambda$test$1(ReloadingDatabasesWhilePerformingGeoLookupsIT.java:102) [internalClusterTest/:?]
	at java.lang.Thread.run(Thread.java:832) [?:?]
```

When deleting a db file during the test and we remove the entry for from the DatabaseRegistry, the entry for that is missing. Looks like a testing issue.

--- a/modules/ingest-geoip/src/internalClusterTest/java/org/elasticsearch/ingest/geoip/ReloadingDatabasesWhilePerformingGeoLookupsIT.java
+++ b/modules/ingest-geoip/src/internalClusterTest/java/org/elasticsearch/ingest/geoip/ReloadingDatabasesWhilePerformingGeoLookupsIT.java
@@ -8,9 +8,11 @@
 
 package org.elasticsearch.ingest.geoip;
 
+import org.apache.lucene.util.LuceneTestCase;
 import org.elasticsearch.client.Client;
 import org.elasticsearch.common.network.InetAddresses;
 import org.elasticsearch.common.util.concurrent.AtomicArray;
+import org.elasticsearch.core.internal.io.IOUtils;
 import org.elasticsearch.index.VersionType;
 import org.elasticsearch.ingest.IngestDocument;
 import org.elasticsearch.ingest.IngestService;
@@ -33,12 +35,14 @@ import java.util.concurrent.atomic.AtomicReference;
 import static org.elasticsearch.ingest.geoip.GeoIpProcessorFactoryTests.copyDatabaseFiles;
 import static org.hamcrest.Matchers.equalTo;
 import static org.hamcrest.Matchers.greaterThan;
+import static org.hamcrest.Matchers.is;
 import static org.hamcrest.Matchers.not;
 import static org.hamcrest.Matchers.notNullValue;
 import static org.hamcrest.Matchers.nullValue;
 import static org.hamcrest.Matchers.sameInstance;
 import static org.mockito.Mockito.mock;
 
+@LuceneTestCase.SuppressFileSystems(value = "ExtrasFS") // Don't randomly add 'extra' files to directory.
 public class ReloadingDatabasesWhilePerformingGeoLookupsIT extends ESTestCase {
 
     /**
@@ -48,10 +52,11 @@ public class ReloadingDatabasesWhilePerformingGeoLookupsIT extends ESTestCase {
      * This failure can be avoided by ensuring that a database is only closed when no
      * geoip processor instance is using the related {@link DatabaseReaderLazyLoader} instance
      */
-    @AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/69980")
     public void test() throws Exception {
+        Path geoIpModulesDir = createTempDir();
+        Path geoIpConfigDir = createTempDir();
         Path geoIpTmpDir = createTempDir();
-        DatabaseRegistry databaseRegistry = createDatabaseRegistry(geoIpTmpDir);
+        DatabaseRegistry databaseRegistry = createRegistry(geoIpModulesDir, geoIpConfigDir, geoIpTmpDir);
         GeoIpProcessor.Factory factory = new GeoIpProcessor.Factory(databaseRegistry);
         Files.copy(LocalDatabases.class.getResourceAsStream("/GeoLite2-City-Test.mmdb"),
             geoIpTmpDir.resolve("GeoLite2-City.mmdb"));
@@ -101,6 +106,13 @@ public class ReloadingDatabasesWhilePerformingGeoLookupsIT extends ESTestCase {
                     DatabaseReaderLazyLoader previous1 = databaseRegistry.get("GeoLite2-City.mmdb");
                     if (Files.exists(geoIpTmpDir.resolve("GeoLite2-City.mmdb"))) {
                         databaseRegistry.removeStaleEntries(List.of("GeoLite2-City.mmdb"));
+                        assertBusy(() -> {
+                            // lazy loader may still be in use by an ingest thread,
+                            // wait for any potential ingest thread to release the lazy loader (DatabaseReaderLazyLoader#postLookup(...)),
+                            // this will do clean it up and actually deleting the underlying file
+                            assertThat(Files.exists(geoIpTmpDir.resolve("GeoLite2-City.mmdb")), is(false));
+                            assertThat(previous1.current(), equalTo(-1));
+                        });
                     } else {
                         Files.copy(LocalDatabases.class.getResourceAsStream("/GeoLite2-City-Test.mmdb"),
                             geoIpTmpDir.resolve("GeoLite2-City.mmdb"), StandardCopyOption.REPLACE_EXISTING);
@@ -146,13 +158,14 @@ public class ReloadingDatabasesWhilePerformingGeoLookupsIT extends ESTestCase {
         for (DatabaseReaderLazyLoader lazyLoader : databaseRegistry.getAllDatabases()) {
             assertThat(lazyLoader.current(), equalTo(0));
         }
+        // Avoid accumulating many temp dirs while running with -Dtests.iters=X
+        IOUtils.rm(geoIpModulesDir, geoIpConfigDir, geoIpTmpDir);
     }
 
-    private static DatabaseRegistry createDatabaseRegistry(Path geoIpTmpDir) throws IOException {
-        Path geoIpDir = createTempDir();
-        copyDatabaseFiles(geoIpDir);
+    private static DatabaseRegistry createRegistry(Path geoIpModulesDir, Path geoIpConfigDir, Path geoIpTmpDir) throws IOException {
+        copyDatabaseFiles(geoIpModulesDir);
         GeoIpCache cache = new GeoIpCache(0);
-        LocalDatabases localDatabases = new LocalDatabases(geoIpDir, createTempDir(), cache);
+        LocalDatabases localDatabases = new LocalDatabases(geoIpModulesDir, geoIpConfigDir, cache);
         DatabaseRegistry databaseRegistry =
             new DatabaseRegistry(geoIpTmpDir, mock(Client.class), cache, localDatabases, Runnable::run);
         databaseRegistry.initialize(mock(ResourceWatcherService.class), mock(IngestService.class));

